opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -43,8 +46,222 @@
 
 namespace Opm
 {
+namespace
+{
+//! \brief A custom deleter for the parallel preconditioners.
+//!
+//! In dune-istl they hold a reference to the sequential preconditioner.
+//! In CPRPreconditioner we use unique_ptr for the memory management.
+//! Ergo we need to construct the sequential preconditioner with new and
+//! make sure that it gets deleted together with the enclosing parallel
+//! preconditioner. Therefore this deleter stores a pointer to it and deletes
+//! it during destruction.
+template<class PREC>
+class ParallelPreconditionerDeleter
+{
+public:
+    ParallelPreconditionerDeleter()
+        : ilu_()
+    {}
+    ParallelPreconditionerDeleter(PREC& ilu)
+    : ilu_(&ilu){}
+    template<class T>
+    void operator()(T* pt)
+    {
+        if(ilu_)
```

Comment:
> [`free(NULL)`] wasn't safe in C89

That depends on what you mean by C89.  Every standard version of C, including the ANSI edition of 1989 which introduced prototypes and `void *` as a generic pointer has sanctioned `free(NULL)` as legal (and a no-op).  Prior to the ANSI standard of 1989 `malloc()` used to return a `char *`.  I don't know enough about `free()` at the time to know if `free(NULL)` behaved in a safe manner.

My point is, however, that `free(NULL)` has been safe in C for more than 25 years and there really should be no need to check for "nullness" in that situation in either C or C++.